compiler/rustc_ast/src/ast_traits.rs

```diff
@@ -0,0 +1,435 @@
+    };
+}
+
+macro_rules! impl_has_attrs_none {
+    ($($T:ty),+ $(,)?) => {
+        $(
+            impl HasAttrs for $T {
+                const SUPPORTS_CUSTOM_INNER_ATTRS: bool = false;
+                fn attrs(&self) -> &[Attribute] {
+                    &[]
+                }
+                fn visit_attrs(&mut self, _f: impl FnOnce(&mut Vec<Attribute>)) {}
+            }
+        )+
+    };
+}
+
+impl_has_attrs!(
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = true,
+    AssocItem,
+    ForeignItem,
+    Item,
+);
+impl_has_attrs!(
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = false,
+    Arm,
+    Crate,
+    Expr,
+    ExprField,
+    FieldDef,
+    GenericParam,
+    Param,
+    PatField,
+    Variant,
+);
+impl_has_attrs_none!(Attribute, AttrItem, Block, Pat, Path, Ty, Visibility);
+
+impl<T: AstDeref<Target: HasAttrs>> HasAttrs for T {
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = T::Target::SUPPORTS_CUSTOM_INNER_ATTRS;
+    fn attrs(&self) -> &[Attribute] {
+        self.ast_deref().attrs()
+    }
+    fn visit_attrs(&mut self, f: impl FnOnce(&mut Vec<Attribute>)) {
+        self.ast_deref_mut().visit_attrs(f)
+    }
+}
+
+impl<T: HasAttrs> HasAttrs for Option<T> {
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = T::SUPPORTS_CUSTOM_INNER_ATTRS;
+    fn attrs(&self) -> &[Attribute] {
+        self.as_ref().map(|inner| inner.attrs()).unwrap_or(&[])
+    }
+    fn visit_attrs(&mut self, f: impl FnOnce(&mut Vec<Attribute>)) {
+        if let Some(inner) = self.as_mut() {
+            inner.visit_attrs(f);
+        }
+    }
+}
+
+impl HasAttrs for StmtKind {
+    // This might be an `StmtKind::Item`, which contains
+    // an item that supports inner attrs
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = true;
+
+    fn attrs(&self) -> &[Attribute] {
+        match self {
+            StmtKind::Local(local) => &local.attrs,
+            StmtKind::Expr(expr) | StmtKind::Semi(expr) => expr.attrs(),
+            StmtKind::Item(item) => item.attrs(),
+            StmtKind::Empty => &[],
+            StmtKind::MacCall(mac) => &mac.attrs,
+        }
+    }
+
+    fn visit_attrs(&mut self, f: impl FnOnce(&mut Vec<Attribute>)) {
+        match self {
+            StmtKind::Local(local) => visit_attrvec(&mut local.attrs, f),
+            StmtKind::Expr(expr) | StmtKind::Semi(expr) => expr.visit_attrs(f),
+            StmtKind::Item(item) => item.visit_attrs(f),
+            StmtKind::Empty => {}
+            StmtKind::MacCall(mac) => visit_attrvec(&mut mac.attrs, f),
+        }
+    }
+}
+
+impl HasAttrs for Stmt {
+    const SUPPORTS_CUSTOM_INNER_ATTRS: bool = StmtKind::SUPPORTS_CUSTOM_INNER_ATTRS;
+    fn attrs(&self) -> &[Attribute] {
+        self.kind.attrs()
+    }
+    fn visit_attrs(&mut self, f: impl FnOnce(&mut Vec<Attribute>)) {
+        self.kind.visit_attrs(f);
+    }
+}
+
+/// Helper trait for the macros below. Abstracts over
+/// the two types of attribute fields that AST nodes
+/// may have (`Vec<Attribute>` or `AttrVec`)
```

Comment:
```suggestion
/// may have (`Vec<Attribute>` or `AttrVec`).
```
Also these comment lines seem too short, this comment could probably fit into two lines.